astropy/utils/data_info.py

```diff
@@ -184,7 +184,82 @@ def _get_data_attribute(dat, attr=None):
     return str(val)
 
 
-class DataInfo:
+class InfoAttribute:
+    def __init__(self, attr, default=None):
+        self.attr = attr
+        self.default = default
+
+    def __get__(self, instance, owner_cls):
+        if instance is None:
+            return self
+
+        return instance._attrs.get(self.attr, self.default)
+
+    def __set__(self, instance, value):
+        if instance is None:
+            # This is an unbound descriptor on the class
+            raise ValueError('cannot set unbound descriptor')
+
+        instance._attrs[self.attr] = value
+
+
+class ParentAttribute:
+    def __init__(self, attr):
+        self.attr = attr
+
+    def __get__(self, instance, owner_cls):
+        if instance is None:
+            return self
+
+        return getattr(instance._parent, self.attr)
+
+    def __set__(self, instance, value):
+        if instance is None:
+            # This is an unbound descriptor on the class
+            raise ValueError('cannot set unbound descriptor')
+
+        setattr(instance._parent, self.attr, value)
+
+
+class DataInfoMeta(type):
+    def __new__(mcls, name, bases, dct):
+        # Ensure that we do not gain a __dict__, which would mean
+        # arbitrary attributes could be set (as a by-product, save
+        # some memory).
+        dct.setdefault('__slots__', [])
+        # Set a default __doc__, mostly to avoid InheritDocstrings
+        # from trying to give us a docstring if none is present.
+        dct.setdefault('__doc__', """\
+Container for meta information like name, description, format.
+This is required when the object is used as a mixin column within
+a table, but can be used as a general way to store meta information.
+""")
+        return super().__new__(mcls, name, bases, dct)
+
+    def __init__(cls, name, bases, dct):
+        super().__init__(name, bases, dct)
+
+        # Define default getters/setters for attributes, if needed.
```

Comment:
I just started looking at this and my immediate question is why it is not possible to use simple class attributes and rely on class-based inheritance.
```
class DataInfo(..):
    name = InfoAttribute()
    dtype = InfoAttribute()
   ...

class BaseColumnInfo(DataInfo):
    parent_table = InfoAttribute(copy=False)  # Attribute knows not to get copied

class ParentDtypeInfo(MixinInfo):
    dtype = ParentAttribute()
```
I ended up going crazy with descriptors for a class that stores most of its attributes in astropy Table meta.  :smile: 
https://github.com/sot/proseco/blob/ad6f861d0fd9eb27cc7034731ed758c9b875f84b/proseco/core.py#L511